src/ripple/basics/algorithm.h

```diff
@@ -0,0 +1,110 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_ALGORITHM_H_INCLUDED
+#define RIPPLE_ALGORITHM_H_INCLUDED
+
+#include <utility>
+
+namespace ripple {
+
+// Requires: [first1, last1) and [first2, last2) are ordered ranges according to
+// comp.
+
+// Effects: For each pair of elements {i, j} in the intersection of the sorted
+// sequences [first1, last1) and [first2, last2), perform action(i, j).
+
+// Note: This algorithm is evolved from std::set_intersection.
+template <class InputIter1, class InputIter2, class Action, class Comp>
+void
+generalized_set_intersection(InputIter1 first1, InputIter1 last1,
+                             InputIter2 first2, InputIter2 last2,
+                             Action action, Comp comp)
+{
+    while (first1 != last1 && first2 != last2)
+    {
+        
+        if (comp(*first1, *first2))       // if *first1 < *first2
+            ++first1;                     //     then reduce first range
+        else
+        {
+            if (!comp(*first2, *first1))  // if *first1 == *first2
+            {                             //     then this is an intersection
+                action(*first1, *first2); //     do the action
+                ++first1;                 //     reduce first range
+            }
+            ++first2;        // Reduce second range because *first2 <= *first1
+        }
+    }
+}
+
+// Requires: [first1, last1) and [first2, last2) are ordered ranges according to
```

Comment:
Should we put in a debug-only assert which uses `std::is_sorted` to verify this precondition?

Reply:
I would only want to put an O(N log N) check in under a system where we had at least a two-level tier of debug checks: basically quick checks and expensive checks.

Rationale:  Today I can run rippled in release mode with asserts on.  But I can't run rippled in debug mode because it can't keep up with the network.  If we make asserts sufficiently expensive, I won't be able to run them (just like debug is today).

Reply:
Thanks for the rationale, @HowardHinnant. Makes sense 👍

Reply:
> But I can't run rippled in debug mode because it can't keep up with the network.

What exactly do you mean by "debug mode" by the way? I'm running a debug build of `rippled` within a gdb session on the main network without issues for months by now to try to catch a rare bug I encountered...

Reply:
Last time I tried running a non-optimized build I had trouble getting and keeping a sync.  Admittedly it has been probably more than a year since I even tried.  I just went to release+asserts and never looked back.

Reply:
I had the same comment in a pending review: 

>It may be a good idea to do:
>
>```
>assert(std::is_sorted(first1, last1));
>assert(std::is_sorted(first2, last2));
>```
>
>Of course, that assumes that those ranges "sort" with the default sorting algorithm which isn't something you can necessarily assume in a "generalized" algorithm.

I think it's reasonable to have the expensive check under an `assert` so that it's optimized away when compiling with `NDEBUG`.

I am generally concerned whenever I see code that's conditionally defined to only execute under debug compilation—`assert` excepted. It's just more that can go wrong and more to screw up. I believe that we should use error handling code for conditions we expect to occur and assertions for situations that should never occur. Passing in an unsorted range seems like the latter.